added CMakeLists.txt for llama2 runner

jeffxtang opened the pull request:
Before `cd executorch` and run `examples/models/llama2/runner_cmake.sh`, you may need to `wget https://download.pytorch.org/libtorch/nightly/cpu/libtorch-shared-with-deps-latest.zip` inside the executorch (or modify the CMakeLists.txt) to avoid a possible error of unable to find c10/macros/cmake_macros.h.

The 3 libs got built: 
$ ls -lt cmake-out/examples/models/llama2
-rw-r--r--   1 jeffxtang  staff   11432 Feb  8 21:27 libtokenizer.a
-rw-r--r--   1 jeffxtang  staff    4936 Feb  8 21:27 libsampler.a
-rw-r--r--   1 jeffxtang  staff   67480 Feb  8 21:27 librunner.a

But there's still some linker error at the end of running the runner_cmake.sh (no more Sampler or Tokenizer related error):
 
<img width="1480" alt="image" src="https://github.com/pytorch/executorch/assets/535090/83176a71-343b-48af-9428-57d241b689a9">

Thoughts on possible fix? @larryliu0820 @lucylq @iseeyuan

jeffxtang committed:
added CMakeLists.txt for llama2 runner

jeffxtang committed:
uncommented cmake_install_executorch_sdk_lib in runner_cmake.sh

larryliu0820 commented:
Thanks for setting this up! For the error you saw, it seems you need to include this subdirectory: https://github.com/pytorch/executorch/blob/main/extension/module/CMakeLists.txt

Also would be good to rely on this util to avoid the churn of managing these source files: https://github.com/pytorch/executorch/blob/main/build/cmake_deps.toml

jeffxtang committed:
updated CMakeLists.txt to incldue extension_module and extension_data_loader

jeffxtang commented:
> Thanks for setting this up! For the error you saw, it seems you need to include this subdirectory: https://github.com/pytorch/executorch/blob/main/extension/module/CMakeLists.txt
> 
> Also would be good to rely on this util to avoid the churn of managing these source files: https://github.com/pytorch/executorch/blob/main/build/cmake_deps.toml

Thanks for the links I updated the CMakeLists.txt (and also added another lib `extension_data_loader`) and can now build the binary but its size is smaller than the binary built with buck and running it just returned "Abort trap: 6".

<img width="600" alt="image" src="https://github.com/pytorch/executorch/assets/535090/f95657fd-732c-4910-996a-28c41edee13b">

Does the CMakeLists.txt look all good? @larryliu0820

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,253 @@
+
+        message(">>>>>> CMAKE_CURRENT_SOURCE_DIR: ${CMAKE_CURRENT_SOURCE_DIR}")
+
+        set(EXPORTED_HEADERS_RUNNER
+            "${CMAKE_CURRENT_SOURCE_DIR}/runner/runner.h"
+            "${CMAKE_CURRENT_SOURCE_DIR}/runner/util.h"
+        )
```
We don't need this because we include the whole executorch/ directory like here https://github.com/pytorch/executorch/blob/main/examples/selective_build/CMakeLists.txt#L45

jeffxtang replied:
without this i got en error
```
Built target flatcc_project
make: *** [all] Error 2
```

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,253 @@
+        )
+
+        # Define a CMake target using runtime.cxx_library
+        add_library(runner${aten_suffix}
+            "${CMAKE_CURRENT_SOURCE_DIR}/runner/runner.cpp"
+            "${CMAKE_CURRENT_SOURCE_DIR}/sampler/sampler.cpp"
+            "${CMAKE_CURRENT_SOURCE_DIR}/tokenizer/tokenizer.cpp"
+        )
```
Would be nice if we can add this buck target into https://github.com/pytorch/executorch/blob/main/build/cmake_deps.toml#L11

and then you can refer to these sources through:
https://github.com/pytorch/executorch/blob/main/CMakeLists.txt#L163-L168

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,253 @@
+
+        # Specify exported dependencies
+        target_link_libraries(runner${aten_suffix} PRIVATE
+            //executorch/backends/xnnpack:xnnpack_backend
+            //executorch/examples/models/llama2/sampler:sampler${aten_suffix}
+            //executorch/examples/models/llama2/tokenizer:tokenizer${aten_suffix}
+            //executorch/examples/models/llama2/runner:runner${aten_suffix}
+            //executorch/extension/evalue_util:print_evalue${aten_suffix}
+            //executorch/extension/runner_util:managed_tensor${aten_suffix}
+            //executorch/extension/module:module${aten_suffix}
+            //executorch/kernels/quantized:generated_lib${aten_suffix}
+            //executorch/kernels/portable:${aten_suffix}generated_lib_all_ops
+            //executorch/runtime/core/exec_aten:lib${aten_suffix}
+            //executorch/runtime/kernel:kernel_includes
```
Do these really work? Seems copied from targets.bzl

jeffxtang replied:
Without these lines, i also got an error:
```
[ 48%] Built target flatcc_project
make: *** [all] Error 2/
```

jeffxtang commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,253 @@
+  ${llama_schema_headers}
+)
+
+function(define_common_targets)
+    foreach(aten IN ITEMS OFF) # ON OFF
```
sure loop removed and $aten_suffix too.

jeffxtang committed:
got rid of for loop of aten and simplified the code

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,226 @@
+  ${llama_schema_headers}
+)
+
+function(define_common_targets)
```
Can we inline this because there's only one call

jeffxtang replied:
yes we can. will push the change soon.

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,226 @@
+)
+
+function(define_common_targets)
+    set(EXPORTED_HEADERS_RUNNER
+        "${CMAKE_CURRENT_SOURCE_DIR}/runner/runner.h"
+        "${CMAKE_CURRENT_SOURCE_DIR}/runner/util.h"
+    )
```
Hmm like I mentioned last time, we don't need to specifically include these headers, if you do this: https://github.com/pytorch/executorch/blob/main/CMakeLists.txt#L250

jeffxtang replied:
tried but got error. not knowing for sure a better way, i put line 108 & 109 directly under `target_sources(runner PRIVATE`.

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,226 @@
+    )
+
+    # Specify exported headers
+    target_include_directories(runner PUBLIC
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../..
+        ${CMAKE_CURRENT_SOURCE_DIR}/runner
+        ${CMAKE_CURRENT_SOURCE_DIR}/sampler
+        ${CMAKE_CURRENT_SOURCE_DIR}/tokenizer
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch/torch/csrc/api/include
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/torch/csrc/api/include/
+    )
```
Same comment as above. As for libtorch include, I don't think we need those?

jeffxtang replied:
Without these, I got error including:
<img width="1399" alt="image" src="https://github.com/pytorch/executorch/assets/535090/a9e1243e-f067-4836-afa8-2ef2f5f92134">

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,226 @@
+        "${CMAKE_CURRENT_SOURCE_DIR}/sampler/sampler.cpp"
+    )
+
+    target_include_directories(sampler PUBLIC
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../..
+        ${CMAKE_CURRENT_SOURCE_DIR}/sampler
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch/torch/csrc/api/include
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/torch/csrc/api/include/
+    )
```
Same comment

larryliu0820 commented on examples/models/llama2/CMakeLists.txt:
```diff
@@ -0,0 +1,226 @@
+        "${CMAKE_CURRENT_SOURCE_DIR}/tokenizer/tokenizer.cpp"
+    )
+
+    target_include_directories(tokenizer PUBLIC
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../..
+        ${CMAKE_CURRENT_SOURCE_DIR}/tokenizer
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch/torch/csrc/api/include
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../third-party/pytorch
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/
+        ${CMAKE_CURRENT_SOURCE_DIR}/../../../libtorch/include/torch/csrc/api/include/
+    )
```
Same

jeffxtang committed:
use inline code to simplify; not specially include the headers

jeffxtang committed:
removed extra target_include_directories
